Tests/Creation/Frame/Content/String/ID3CreditsListFrameCreatorTest.swift

```diff
@@ -17,7 +17,7 @@ class ID3CreditsListFrameCreatorTest: XCTestCase {
                 synchsafeEncoder: SynchsafeIntegerEncoder()
             ),
             frameFlagsCreator: ID3FrameFlagsCreator(),
-            stringToBytesAdapter: ID3UTF16StringToByteAdapter(
+            stringToBytesAdapter: ID3ISO88591StringToByteAdapter(
```

Comment:
I changed the StringToByteAdapter, and now the failure message is:

```
XCTAssertEqual failed: ("[34, 0, 0, 0, 25, 0, 0, 0, 109, 117, 115, 105, 99, 105, 97, 110, 0, 0, 77, 117, 115, 105, 99, 105, 97, 110, 32, 78, 97, 109, 101, 0]") 
is not equal to ("[34, 0, 0, 0, 52, 0, 0, 1, 109, 117, 115, 105, 99, 105, 97, 110, 0, 77, 117, 115, 105, 99, 105, 97, 110, 32, 78, 97, 109, 101, 0]")```

So it's expecting another null byte between the strings?

Which is fine, I can rewrite the test to include that, but I'd like to understand why so I'm not just rigging the test to pass without making sure I'm actually doing it right.

Reply:
I don’t think there should be two 0 bytes unless the encoding is 16‐bit. (In 8‐bit, null is `0x00`; in 16‐bit, it is `0x0000`.) Without looking at the code, there are (at least) two possibilities:

1. The string encoding method is terminating the string with null automatically, and then you terminating it with another null manually.
2. The string encoding method has its wires crossed and is encoding the string in 8‐bit, but the termination null in 16‐bit.

Reply:
For the most part, I'm imitating what you did with the UserDefinedTextInformation frame, or the CommentTypes frame since you used that type for the UserText frame in some places.

Where would I look to see if this is happening?

This is the CreditsListFrameCreator:

```swift
class ID3CreditsListFrameCreator: CreditsListFrameCreator {
    
    private let frameContentSizeCalculator: FrameContentSizeCalculator
    private let frameFlagsCreator: FrameFlagsCreator
    private let stringToBytesAdapter: StringToBytesAdapter
    
    init(frameContentSizeCalculator: FrameContentSizeCalculator,
         frameFlagsCreator: FrameFlagsCreator,
         stringToBytesAdapter: StringToBytesAdapter) {
        self.frameContentSizeCalculator = frameContentSizeCalculator
        self.frameFlagsCreator = frameFlagsCreator
        self.stringToBytesAdapter = stringToBytesAdapter
    }
    
    func createFrame(
        frameIdentifier: [UInt8],
        version: ID3Version,
        role: String,
        person: String
    ) -> [UInt8] {
        let frameContents = [UInt8](
            [
                stringToBytesAdapter.adapt(
                    string: role,
                    for: version,
                    includingEncoding: true,
                    includingTermination: true
                ),
                stringToBytesAdapter.adapt(
                    string: person,
                    for: version,
                    includingEncoding: true,
                    includingTermination: true
                )
                ].joined()
        )
        
        return [UInt8](
            [
                frameIdentifier,
                frameContentSizeCalculator.calculateSizeOf(
                    content: frameContents, version: version),
                frameFlagsCreator.createFor(version: version),
                frameContents
                ].joined()
        )
    }
}
```

So it does include the termination on both strings, which is from your (changes to?) the StringToBytesAdapter. Since this is only testing the FrameCreator, I don't think anything I included in the Parsing (which I'm a lot iffier on) could be affecting it, could it?

Everything I can see for the encoding is ISO8859-1. It's using `UInt32ToByteArrayAdapterUsingUnsafePointer` like the test I am emulating. Could that be an issue?

Reply:
Actually the first of the pair of nulls is the termination of the preceding string, and the second (wrongly) is the encoding of the following string (Latin 1). Only the very first string in the list should include the encoding specifier. The rest should use `includingEncoding: false`.

Reply:
Ah ok. Thank you!

Reply:
Okay, this next part is the tricky bit. Per the spec, I've got the `role` and `person` strings in place, each terminated.

Now I need to be able to add more `role`:`person` pairs *on the same frame*, because the spec (at least for `.version3`, says there should only be one IPLS frame per tag.

So, on the API end, ideally I think it should be just like adding a new frame, but behind the scenes, what is actually happening is that the strings are being appended to the existing frame, such that the frame reads `role1`,`person1`,`role2`,`person2`,`role3`,`person3, person4, person5` (since the `person` string should be able to be a comma delimited string with multiple people fulfilling the same role, per the `.version4` spec.)

And on the parsing end, it should basically come back as though there were multiple `IPLS/TIPL/TMCL` frames rather than just one, with each `role`:`person` pair being returned separately.

So this is different from, for example, the `TXXX` frame where there is more than one frame allowed per tag as long as each one has a different description (which is still an issue I need to tackle in `ID3TagEditor`.)

I suspect in the end the code isn't going to be that complicated, but I have been looking at it since last night scratching my head because I can't figure out where this functionality should go? I'm just drawing a blank on how to get started with it.

Reply:
I'd ask if maybe the `role:person` pair should be handled as a dictionary entry instead of separate strings, but that is sort of made impossible by the fact that there may be multiple people fulfilling the same role.

Reply:
Start by making the API that just does exactly like the specification says and make sure it works. (Then you can add your own convenience methods on top of that.)

That means, for starters, instead of these parameters...

```swift
role: String,
person: String
```

...you need this...

```swift
entries: [(role: String, person: String)]
```

Reply:
Thank you! Will do!